api/layer/multipart_upload.go

```diff
@@ -0,0 +1,459 @@
+	UploadHasACL                  = "S3-Upload-Has-ACL"
+	UploadPartKeyPrefix           = ".upload-"
+
+	UploadMinSize = 5 * 1048576
+)
+
+type (
+	UploadInfoParams struct {
+		UploadID string
+		Bkt      *data.BucketInfo
+		Key      string
+	}
+
+	UploadPartParams struct {
+		Info       *UploadInfoParams
+		PartNumber int
+		Size       int64
+		Reader     io.Reader
+		Header     map[string]string
+	}
+
+	CompleteMultipartParams struct {
+		Info  *UploadInfoParams
+		Parts []*Part
+	}
+
+	Part struct {
+		ETag       string
+		PartNumber int
+	}
+
+	ListMultipartUploadsParams struct {
+		Bkt            *data.BucketInfo
+		Delimiter      string
+		EncodingType   string
+		KeyMarker      string
+		MaxUploads     int
+		Prefix         string
+		UploadIdMarker string
+	}
+
+	ListPartsParams struct {
+		Info             *UploadInfoParams
+		MaxParts         int
+		PartNumberMarker int
+	}
+
+	ListPartsInfo struct {
+		Parts                []*data.ObjectInfo
+		Owner                *owner.ID
+		NextPartNumberMarker int
+		IsTruncated          bool
+	}
+
+	ListMultipartUploadsInfo struct {
+		Prefixes           []string
+		Uploads            []*UploadInfo
+		IsTruncated        bool
+		NextKeyMarker      string
+		NextUploadIdMarker string
+	}
+	UploadInfo struct {
+		IsDir    bool
+		Key      string
+		UploadID string
+		Owner    *owner.ID
+		Created  time.Time
+	}
+)
+
+func (n *layer) UploadPart(ctx context.Context, p *UploadPartParams) (*data.ObjectInfo, error) {
+	if p.PartNumber != 0 {
+		_, err := n.GetUploadInitInfo(ctx, p.Info)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if p.Header == nil {
+		p.Header = make(map[string]string)
+	}
+
+	p.Header[UploadIDAttributeName] = p.Info.UploadID
+	p.Header[UploadPartNumberAttributeName] = strconv.Itoa(p.PartNumber)
+	p.Header[UploadKeyAttributeName] = p.Info.Key
+	p.Header[attrVersionsIgnore] = "true"
+
+	params := &PutObjectParams{
+		Bucket: p.Info.Bkt.Name,
+		Object: createUploadPartName(p.Info.UploadID, p.Info.Key, p.PartNumber),
+		Size:   p.Size,
+		Reader: p.Reader,
+		Header: p.Header,
+	}
+
+	return n.objectPut(ctx, p.Info.Bkt, params)
+}
+
+func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipartParams) (*data.ObjectInfo, error) {
+	var (
+		obj             *data.ObjectInfo
+		uploadCompleted = false
+	)
+
+	for i := 1; i < len(p.Parts); i++ {
+		if p.Parts[i].PartNumber < p.Parts[i-1].PartNumber {
+			return nil, errors.GetAPIError(errors.ErrInvalidPartOrder)
+		}
+	}
+
+	objects, err := n.getUploadParts(ctx, p.Info.UploadID, p.Info.Key, p.Info.Bkt)
+	if err != nil {
+		if err == errors.GetAPIError(errors.ErrNoSuchUpload) {
+			return n.headLastVersionIfNotDeleted(ctx, p.Info.Bkt, p.Info.Key)
+		}
+		return nil, err
+
+	}
+
+	if len(objects) < len(p.Parts) {
+		return nil, errors.GetAPIError(errors.ErrInvalidPart)
+	}
+
+	parts := make([]*data.ObjectInfo, 0, len(p.Parts))
+
+	for i, part := range p.Parts {
+		info := objects[part.PartNumber]
```

Comment:
As I understand this can cause panic because `part.PartNumber` can be greater than `len(objects)`. https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html. For example we can upload parts with number 4, 5, 15 and then complete multipart upload. So we have 3 object but the part numbers are greater than 3.